### PR TITLE
[WIP] Add verbose to NCA and MLKR

### LINKED ISSUE #67: Addition of verbosity setting for all of the algorithms
Currently, none of the algorithms have a verbosity setting while fitting. It is useful to see some details when the fitting process is going on.

### PR DESCRIPTION
**Pull request by wdevazelhes**
Fixes #67.

This PR adds a verbose argument for `NCA` and `MLKR`. Here is an example on a snippet. 

```python
from sklearn.datasets import load_digits
from metric_learn import NCA
dataset = load_digits()
X, y = dataset.data, dataset.target
nca = NCA(verbose=True, max_iter=5)
nca.fit(X, y)
```
Returns: 

```pytb
[NCA]
[NCA]  Iteration      Objective Value    Time(s)
[NCA] ------------------------------------------
[NCA]          0         1.436318e+03       0.23
[NCA]          1         1.770346e+03       0.24
[NCA]          2         1.777018e+03       0.26
[NCA]          3         1.779409e+03       0.25
[NCA]          4         1.781304e+03       0.26
[NCA]          5         1.782404e+03       0.25
/home/will/Code/metric-learn-bis/metric_learn/nca.py:110: ConvergenceWarning: [NCA] NCA did not converge: b'STOP: TOTAL NO. of ITERATIONS REACHED LIMIT'
  cls_name, opt_result.message), ConvergenceWarning)
[NCA] Training took 1533143523.76s.
```

The code mostly comes from scikit-learn's NCA PR (https://github.com/scikit-learn/scikit-learn/pull/10058), and this verbose code in particular was taken mostly from scikit-learn's LMNN PR (https://github.com/scikit-learn/scikit-learn/pull/8602)

### TODO: 
- [ ] Fix the convergence warning test for `MLKR` (related to #104)

### CODE REVIEW

File: metric_learn/mlkr.py [hunk 1]
```diff
@@ -79,34 +87,68 @@ def fit(self, X, y):
...
-  np.fill_diagonal(K, 1)
-  W = 2 * K * (np.outer(ydiff, ydiff) / denom)
-  # note: this is the part that the matlab impl drops to C for
-  M = (dX.T * W.ravel()).dot(dX)
-  grad = 2 * A.dot(M)
-  return cost, grad.ravel()
+  def _loss(self, flatA, X, y, dX):
+
+    if self.n_iter_ == 0:
+      if self.verbose:
```
Comment (perimosocordiae) [inline metric_learn/mlkr.py]: I'd combine these into a single condition.

File: metric_learn/mlkr.py [hunk 2]
```diff
@@ -79,34 +87,68 @@ def fit(self, X, y):
...
+        header_fields = ['Iteration', 'Objective Value', 'Time(s)']
+        header_fmt = '{:>10} {:>20} {:>10}'
+        header = header_fmt.format(*header_fields)
+        cls_name = self.__class__.__name__
+        print('[{}]'.format(cls_name))
+        print('[{}] {}\n[{}] {}'.format(cls_name, header, cls_name,
+                                        '-' * len(header)))
```
Comment (perimosocordiae) [inline metric_learn/mlkr.py]: Named format slots might make this easier to read. That is, like the "by name" examples here: https://docs.python.org/2/library/string.html#format-examples

File: metric_learn/mlkr.py [hunk 3]
```diff
@@ -79,34 +87,68 @@ def fit(self, X, y):
...
+
+    t_funcall = time.time()
```
Comment (perimosocordiae) [inline metric_learn/mlkr.py]: I'm not a big fan of the `t_` prefix for times. How about `start_time`?

File: metric_learn/mlkr.py (hunk 2 above)
Comment (wdevazelhes) [reply metric_learn/mlkr.py]: Agreed, done (except for the header field because there I thought keyword arguments would be more redundant than really helping reading , but tell me if you don't think so)

File: metric_learn/mlkr.py (hunk 3 above)
Comment (wdevazelhes) [reply metric_learn/mlkr.py]: I agree, done

**Comment by wdevazelhes** [comment]
I just noticed that the Conjugate Gradient method (used in `MLKR`) has a different way to count iterations: unlike `L-BFGS-B`, one iteration is not equal to one function call, so the way to print iteration number, objective and gradient, as done in this PR does not work for `MLKR`. 

I was thinking I could maybe instead print the cost function for each function evaluation instead of each iteration, for `MLKR` ? 

Or we could use `L-BFGS-B` for `MLKR` too ? @perimosocordiae was there a particular reason to use Conjugate Gradient ? 
Looking into the documentation of  Conjugate Gradient https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.fmin_cg.html, at `Notes` they say that it works better if the function has a global minima, which I don't think there is, right ? (any rotation of an optimal transformation `A` would be optimal too because conserving distances)
@bellet any thought on that ? 

Note: here is a thread that compares the two methods: https://scicomp.stackexchange.com/questions/507/bfgs-vs-conjugate-gradient-method

**Comment by perimosocordiae** [comment]
The [MLKR paper](http://proceedings.mlr.press/v2/weinberger07a/weinberger07a.pdf) says this about the optimization problem, after briefly describing a gradient descent algorithm:

> Of course, other methods for minimizing (3) such as conjugate gradient, stochastic gradient or BFGS may also be used and might lead to faster convergence results.

The initial MLKR implementation by @dsquareindia used a hand-rolled gradient descent, and I later converted this to use `scipy.optimize.minimize` probably without thinking too hard about the choice of method.

So in summary: if you think L_BFGS-B will work better, feel free to try it out! :+1:

**Comment by bellet** [comment]
L-BFGS-B should also work well (maybe slightly better) at the cost of slightly higher memory. Maybe you can indeed give it a try, this would make it more consistent with NCA

**Comment by wdevazelhes** [comment]
Allright, we can go for L-BFGS-B for now (it can easily be changed in the future)
I just added a commit which also uses less features for the test of `make_regression`, making the task easier to train on hence making MLKR do really 2 iterations (and hitting max_iter in the test), and not stopping weirdly at the beginning (see https://github.com/metric-learn/metric-learn/issues/104)

**Comment by perimosocordiae** [comment]
Changing the underlying optimizer is a bit of a compatibility break, too, so I'm fine with making it explicit by changing the keyword arguments.

**Comment by wdevazelhes** [comment]
I just realized in fact `L-BFGS-B` also has an `epsilon` parameter, so maybe we want to keep it ? And also add it to `NCA` consequently ?

**Comment by bellet** [comment]
From my understanding, `epsilon` is useless here because you already provide a function to compute the exact gradient (this is also true for NCA)

**Comment by wdevazelhes** [comment]
Oh yes that's right
I'll provide a deprecation warning similar to the one in NCA and I think we'll be good to merge

**Comment by wdevazelhes** [comment]
I just realized that `MLKR` is introduced in master (will be introduced in the next release), so I guess actually we don't need any deprecation ?

**Comment by bellet** [comment]
Makes sense indeed :-) But need to fix the travis failure

**Comment by wdevazelhes** [comment]
Done
